src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -633,17 +724,56 @@ private static Expression BindParameterFromRouteValueOrQueryString(ParameterInfo
             return BindParameterFromValue(parameter, Expression.Coalesce(routeValue, queryValue), factoryContext);
         }
 
-        private static Expression BindParameterFromBody(Type parameterType, bool allowEmpty, FactoryContext factoryContext)
+        private static Expression BindParameterFromBody(ParameterInfo parameter, bool allowEmpty, FactoryContext factoryContext)
         {
             if (factoryContext.JsonRequestBodyType is not null)
             {
                 throw new InvalidOperationException("Action cannot have more than one FromBody attribute.");
             }
 
-            factoryContext.JsonRequestBodyType = parameterType;
-            factoryContext.AllowEmptyRequestBody = allowEmpty;
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
+            factoryContext.JsonRequestBodyType = parameter.ParameterType;
+            factoryContext.AllowEmptyRequestBody = allowEmpty || isOptional;
```

Comment:
If someone explicitly sets `EmptyBodyBehavior.Disallow`, we probably shouldn't let the nullability of the parameter override that. I wonder if we should change `IFromBodyMetadata.AllowEmpty` from `bool` to `bool?`.

Reply:
Yeah, we currently make the assumption that `EmptyBodyBehavior.Disallow` and `EmptyBodyBehavior.Default` both mean that `allowEmpty = false` in our `FromBody` attribute.

This gets it annoying territory from a code-clean perspective because we're trying to map a ternary value into a binary value, so we're bound to lose some fidelity. Making `AllowEmpty` a nullable value might help us maintain the ternary mapping though AKA

* EmptyBodyBehavior.Default == null
* EmptyBodyBehavior.Disallow = false
* EmptyBodyBehavior.Allow == true

I filed https://github.com/dotnet/aspnetcore/issues/34567 as a follow-up since this will probably require a public API change.